File: hack/fixtures/raw/raw.go
// Command raw captures an unconverted usage response for fixture development.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"time"
)

var (
	versionPattern     = regexp.MustCompile(`\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?`)
	unsafeFilenameChar = regexp.MustCompile(`[^A-Za-z0-9.-]+`)
	httpClient         = &http.Client{Timeout: 15 * time.Second}
)

func main() {
	if len(os.Args) != 2 {
		fatalf("usage: go run ./hack/fixtures/raw <codex|claude|antigravity|all>")
	}

	providers := []string{os.Args[1]}
	if os.Args[1] == "all" {
		providers = []string{"codex", "claude", "antigravity"}
	}

	for _, provider := range providers {
		if err := capture(provider); err != nil {
			fatalf("%s: %v", provider, err)
		}
	}
}

func capture(provider string) error {
	var cliName string
	switch provider {
	case "codex", "claude":
		cliName = provider
	case "antigravity":
		cliName = "agy"
	default:
		return fmt.Errorf("unknown provider %q", provider)
	}

	version, err := cliVersion(cliName)
	if err != nil {
		return err
	}

	var raw []byte
	switch provider {
	case "codex":
		raw, err = fetchCodex()
	case "claude":
		raw, err = fetchClaude()
	case "antigravity":
		raw, err = commandOutput("agy", "-p", "/usage", "--output-format", "json")
	}
	if err != nil {
		return err
	}

	formatted, err := formatJSON(raw)
	if err != nil {
		return fmt.Errorf("invalid JSON response: %w", err)
	}

	var planType string
	if provider == "codex" {
		formatted, err = redactCodex(formatted)
		if err != nil {
			return err
		}
		planType, err = codexPlanType(formatted)
		if err != nil {
			return err
		}
	}

	_, sourceFile, _, ok := runtime.Caller(0)
	if !ok {
		return fmt.Errorf("could not locate output directory")
	}
	var filename string
	if provider == "codex" {
		filename = fmt.Sprintf("%s-%s_%s_%s.json", provider, version, planType, time.Now().Format("2006-01-02"))
	} else {
		filename = fmt.Sprintf("%s-%s_%s.json", provider, version, time.Now().Format("2006-01-02"))
	}
	outputPath := filepath.Join(filepath.Dir(sourceFile), filename)
	if err := os.WriteFile(outputPath, formatted, 0o644); err != nil {
		return fmt.Errorf("write %s: %w", outputPath, err)
	}

	fmt.Println("Wrote", outputPath)
	return nil
}

func cliVersion(name string) (string, error) {
	output, err := commandOutput(name, "--version")
	if err != nil {
		return "", err
	}
	version := versionPattern.FindString(string(output))
	if version == "" {
		return "", fmt.Errorf("could not determine %s version from %q", name, strings.TrimSpace(string(output)))
	}
	return version, nil
}

func commandOutput(name string, args ...string) ([]byte, error) {
	command := exec.Command(name, args...)
	output, err := command.Output()
	if err == nil {
		return output, nil
	}
	if exitError, ok := err.(*exec.ExitError); ok && len(exitError.Stderr) != 0 {
		return nil, fmt.Errorf("%s: %s", name, strings.TrimSpace(string(exitError.Stderr)))
	}
	return nil, fmt.Errorf("%s: %w", name, err)
}

func fetchCodex() ([]byte, error) {
	var credentials struct {
		Tokens struct {
			AccessToken string `json:"access_token"`
		} `json:"tokens"`
	}
	if err := readCredentials(filepath.Join(".codex", "auth.json"), &credentials); err != nil {
		return nil, err
	}
	if credentials.Tokens.AccessToken == "" {
		return nil, fmt.Errorf("Codex credentials contain no access token")
	}
	return fetch("https://chatgpt.com/backend-api/wham/usage", map[string]string{
		"Authorization": "Bearer " + credentials.Tokens.AccessToken,
	})
}

func fetchClaude() ([]byte, error) {
	var credentials struct {
		ClaudeAIOAuth struct {
			AccessToken string `json:"accessToken"`
		} `json:"claudeAiOauth"`
	}
	if err := readCredentials(filepath.Join(".claude", ".credentials.json"), &credentials); err != nil {
		return nil, err
	}
	if credentials.ClaudeAIOAuth.AccessToken == "" {
		return nil, fmt.Errorf("Claude credentials contain no access token")
	}
	return fetch("https://api.anthropic.com/api/oauth/usage", map[string]string{
		"Authorization":  "Bearer " + credentials.ClaudeAIOAuth.AccessToken,
		"anthropic-beta": "oauth-2025-04-20",
	})
}

func readCredentials(relativePath string, target any) error {
	home, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("find home directory: %w", err)
	}
	path := filepath.Join(home, relativePath)
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("read %s: %w", path, err)
	}
	if err := json.Unmarshal(data, target); err != nil {
		return fmt.Errorf("parse %s: %w", path, err)
	}
	return nil
}

func fetch(url string, headers map[string]string) ([]byte, error) {
	request, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	for name, value := range headers {
		request.Header.Set(name, value)
	}

	response, err := httpClient.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("usage request failed: %s", response.Status)
	}
	return io.ReadAll(response.Body)
}

func formatJSON(raw []byte) ([]byte, error) {
	var formatted bytes.Buffer
	if err := json.Indent(&formatted, bytes.TrimSpace(raw), "", "  "); err != nil {
		return nil, err
	}
	formatted.WriteByte('\n')
	return formatted.Bytes(), nil
}

func codexPlanType(data []byte) (string, error) {
	var response struct {
		PlanType string `json:"plan_type"`
	}
	if err := json.Unmarshal(data, &response); err != nil {
		return "", fmt.Errorf("parse plan_type: %w", err)
	}
	if response.PlanType == "" {
		return "unknown", nil
	}
	// plan_type is interpolated straight into the output filename below, so
	// strip anything that isn't a safe filename character (e.g. a path
	// separator) rather than trusting the API response as-is.
	sanitized := unsafeFilenameChar.ReplaceAllString(response.PlanType, "-")
	if sanitized == "" {
		return "unknown", nil
	}
	return sanitized, nil
}

func redactCodex(data []byte) ([]byte, error) {
	redactions := map[string]string{
		"user_id": "user_abc",
		"email":   "user@example.com",
	}
	for field, replacement := range redactions {
		pattern := regexp.MustCompile(`(?m)^(\s*"` + regexp.QuoteMeta(field) + `"\s*:\s*)"[^"]*"`)
		if !pattern.Match(data) {
			return nil, fmt.Errorf("Codex response contains no %q field to redact", field)
		}
		data = pattern.ReplaceAll(data, []byte(`${1}"`+replacement+`"`))
	}
	return data, nil
}

func fatalf(format string, args ...any) {
	fmt.Fprintln(os.Stderr, "raw:", fmt.Sprintf(format, args...))
	os.Exit(1)
}

File: hack/fixtures/raw.mjs
#!/usr/bin/env node

// Command raw captures an unconverted usage response for fixture development.
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import { CODEX_AUTH_RELATIVE_PATH, CLAUDE_CREDENTIALS_RELATIVE_PATH } from '../lib/credential-paths.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const VERSION_PATTERN = /\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?/;
const UNSAFE_FILENAME_CHAR = /[^A-Za-z0-9.-]+/g;
const TIMEOUT_MS = 15000;

function fatalf(msg) {
  console.error(`raw: ${msg}`);
  process.exit(1);
}

function commandOutput(name, args = []) {
  const res = spawnSync(name, args, { encoding: 'utf8' });
  if (res.error) {
    throw new Error(`${name}: ${res.error.message}`);
  }
  if (res.status !== 0) {
    const err = res.stderr && res.stderr.trim() ? res.stderr.trim() : `exit code ${res.status}`;
    throw new Error(`${name}: ${err}`);
  }
  return res.stdout;
}

function cliVersion(name) {
  const output = commandOutput(name, ['--version']);
  const match = output.match(VERSION_PATTERN);
  if (!match) {
    throw new Error(`could not determine ${name} version from "${output.trim()}"`);
  }
  return match[0];
}

function readCredentials(relativePath) {
  const fullPath = path.join(os.homedir(), relativePath);
  if (!fs.existsSync(fullPath)) {
    throw new Error(`read ${fullPath}: file not found`);
  }
  const content = fs.readFileSync(fullPath, 'utf8');
  try {
    return JSON.parse(content);
  } catch (err) {
    throw new Error(`parse ${fullPath}: ${err.message}`);
  }
}

async function fetchJSON(url, headers) {
  const res = await fetch(url, {
    headers,
    signal: AbortSignal.timeout(TIMEOUT_MS),
  });
  if (!res.ok) {
    throw new Error(`usage request failed: ${res.status} ${res.statusText}`);
  }
  return res.json();
}

async function fetchCodex() {
  const creds = readCredentials(CODEX_AUTH_RELATIVE_PATH);
  const token = creds?.tokens?.access_token;
  if (!token) {
    throw new Error('Codex credentials contain no access token');
  }
  return fetchJSON('https://chatgpt.com/backend-api/wham/usage', {
    Authorization: `Bearer ${token}`,
  });
}

async function fetchClaude() {
  const creds = readCredentials(CLAUDE_CREDENTIALS_RELATIVE_PATH);
  const token = creds?.claudeAiOauth?.accessToken;
  if (!token) {
    throw new Error('Claude credentials contain no access token');
  }
  return fetchJSON('https://api.anthropic.com/api/oauth/usage', {
    Authorization: `Bearer ${token}`,
    'anthropic-beta': 'oauth-2025-04-20',
  });
}

function fetchAntigravity() {
  const raw = commandOutput('agy', ['-p', '/usage', '--output-format', 'json']);
  try {
    return JSON.parse(raw);
  } catch (err) {
    throw new Error(`invalid JSON response: ${err.message}`);
  }
}

function formatDate(date = new Date()) {
  return date.toISOString().slice(0, 10);
}

function codexPlanType(data) {
  const plan = data?.plan_type;
  if (!plan) return 'unknown';
  const sanitized = plan.replace(UNSAFE_FILENAME_CHAR, '-');
  return sanitized || 'unknown';
}

function redactCodex(data) {
  if (!('user_id' in data) || !('email' in data)) {
    throw new Error('Codex response contains no user_id or email field to redact');
  }
  data.user_id = 'user_abc';
  data.email = 'user@example.com';
}

const PROVIDERS = {
  codex: { cli: 'codex', fetch: fetchCodex },
  claude: { cli: 'claude', fetch: fetchClaude },
  antigravity: { cli: 'agy', fetch: fetchAntigravity },
};

async function capture(provider) {
  const config = PROVIDERS[provider];
  if (!config) {
    throw new Error(`unknown provider "${provider}"`);
  }

  const version = cliVersion(config.cli);
  const data = await config.fetch();

  let filename;
  const dateStr = formatDate();
  if (provider === 'codex') {
    redactCodex(data);
    const planType = codexPlanType(data);
    filename = `${provider}-${version}_${planType}_${dateStr}.json`;
  } else {
    filename = `${provider}-${version}_${dateStr}.json`;
  }

  const outputDir = path.join(__dirname, 'raw');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const outputPath = path.join(outputDir, filename);
  fs.writeFileSync(outputPath, JSON.stringify(data, null, 2) + '\n', 'utf8');
  console.log(`Wrote ${outputPath}`);
}

async function main() {
  const target = process.argv[2];
  const validProviders = ['codex', 'claude', 'antigravity', 'all'];

  if (!target || !validProviders.includes(target) || process.argv.length !== 3) {
    fatalf('usage: node ./hack/fixtures/raw.mjs <codex|claude|antigravity|all>');
  }

  const providers = target === 'all' ? ['codex', 'claude', 'antigravity'] : [target];

  for (const provider of providers) {
    try {
      await capture(provider);
    } catch (err) {
      fatalf(`${provider}: ${err.message}`);
    }
  }
}

main();
